crates/noirc_evaluator/src/errors.rs

```diff
@@ -1,151 +1,101 @@
+//!
+//! An Error of the former is a user Error
+//!
+//! An Error of the latter is an error in the implementation of the compiler
+use acvm::FieldElement;
 use noirc_errors::{CustomDiagnostic as Diagnostic, FileDiagnostic, Location};
 use thiserror::Error;
 
-#[derive(Debug)]
-pub struct RuntimeError {
-    pub location: Option<Location>,
-    pub kind: RuntimeErrorKind,
-}
-
-impl RuntimeError {
-    // XXX: In some places, we strip the span because we do not want span to
-    // be introduced into the binary op or low level function code, for simplicity.
-    //
-    // It's possible to have it there, but it means we will need to proliferate the code with span
-    //
-    // This does make error reporting, less specific!
-    pub fn remove_span(self) -> RuntimeErrorKind {
-        self.kind
-    }
-
-    pub fn new(kind: RuntimeErrorKind, location: Option<Location>) -> RuntimeError {
-        RuntimeError { location, kind }
-    }
-
-    // Keep one of the two location which is Some, if possible
-    // This is used when we optimize instructions so that we do not lose track of location
-    pub fn merge_location(a: Option<Location>, b: Option<Location>) -> Option<Location> {
-        match (a, b) {
-            (Some(loc), _) | (_, Some(loc)) => Some(loc),
-            (None, None) => None,
-        }
-    }
+#[derive(Debug, PartialEq, Eq, Clone, Error)]
+pub enum RuntimeError {
+    // We avoid showing the actual lhs and rhs since most of the time they are just 0
+    // and 1 respectively. This would confuse users if a constraint such as
+    // assert(foo < bar) fails with "failed constraint: 0 = 1."
+    #[error("Failed constraint")]
+    FailedConstraint { lhs: FieldElement, rhs: FieldElement, location: Option<Location> },
+    #[error(transparent)]
+    ICEError(#[from] ICEError),
+    #[error("Index out of bounds, array has size {index:?}, but index was {array_size:?}")]
+    IndexOutOfBounds { index: usize, array_size: usize, location: Option<Location> },
+    #[error("All Witnesses are by default u{num_bits:?} Applying this type does not apply any constraints.\n We also currently do not allow integers of size more than {num_bits:?}, this will be handled by BigIntegers.")]
+    InvalidRangeConstraint { num_bits: u32, location: Option<Location> },
+    #[error("Expected array index to fit into a u64")]
+    TypeConversion { from: String, into: String, location: Option<Location> },
+    #[error("{name:?} is not initialized")]
+    UnInitialized { name: String, location: Option<Location> },
+    #[error("Integer sized {num_bits:?} is over the max supported size of {max_num_bits:?}")]
+    UnsupportedIntegerSize { num_bits: u32, max_num_bits: u32, location: Option<Location> },
 }
 
-impl From<RuntimeErrorKind> for RuntimeError {
-    fn from(kind: RuntimeErrorKind) -> RuntimeError {
-        RuntimeError { location: None, kind }
-    }
+#[derive(Debug, PartialEq, Eq, Clone, Error)]
+pub enum ICEError {
```

Comment:
Internal Compiler Error Error?

Maybe `InternalError`. Although the exact boundary between this and RuntimeError is a tad unclear.